### ISSUE TITLE
Fatal error: Allowed memory size

### ISSUE THREAD

**Issue by jenswittmann** [Jako/PDFResource#18]
Hello Jako,
thanks for your fast help! For my current project i get the following error:
`Fatal error: Allowed memory size of 134217728 bytes exhausted (tried to allocate 262144 bytes) in core/xpdo/om/xpdoobject.class.php on line 1737`

The template is empty and there are now tables in it.

Thanks a lot!
Jens

**Comment by pashajanx** [comment]
Have you solved the problem?

**Comment by jenswittmann** [comment]
No, not now. Do you got the same error?

**Comment by pashajanx** [comment]
Yes

**Comment by Jako** [comment]
It could be some recursion issue. But thats difficult to locate.

**Comment by Jako** [comment]
You could try to change this line https://github.com/Jako/PDFResource/blob/master/core/components/pdfresource/model/pdfresource/pdfresource.class.php#L193 to 

```
$tvs = $resource->getTemplateVars();
```

**Comment by Jako** [comment]
Or check with some `die('until here');` the line where the error occurs first in the createPDF method of the PDFResource class.

**Comment by pashajanx** [comment]
An error occurs in the getOption class

**Comment by Jako** [comment]
@pashajanx Sorry, but that answer does not really help. Could you see where the getOption method was called from before the error occurs?

**Comment by pashajanx** [comment]
@Jako It seems to me from the plugin.
Commented out the line:
`$modx->loadClass($className, $pdfresource->getOption('modelPath') . 'pdfresource/events/', true, true);`
Error gone missing.

**Comment by pashajanx** [comment]
Found what causes the error
```
$c = $modx->newQuery('modTemplateVarTemplate');
$c->leftJoin('modTemplateVar', 'modTemplateVar', array('modTemplateVarTemplate.tmplvarid = modTemplateVar.id'));
$c->select($modx->getSelectColumns('modTemplateVar', 'modTemplateVar', '', array('name')));
$c->select($modx->getSelectColumns('modTemplateVarTemplate', 'modTemplateVarTemplate', '', array('templateid')));
$c->where(array(
            'modTemplateVar.name' => 'create_pdf',
            'modTemplateVarTemplate.templateid' => '12'
        ));
$assigned = $modx->getObject('modTemplateVarTemplate', $c);
print_r($assigned);
```
The console showed an error:
`Fatal error: Allowed memory size of 268435456 bytes exhausted (tried to allocate 130968 bytes) in .../core/xpdo/om/xpdoobject.class.php on line 1737`

**Comment by Jako** [comment]
Have you done the patch in comment https://github.com/Jako/PDFResource/issues/18#issuecomment-348756167 ?

**Comment by pashajanx** [comment]
Yes, the error has remained.

**Comment by pashajanx** [comment]
Replaced getobject with getIterator. Has earned
https://github.com/Jako/PDFResource/blob/master/core/components/pdfresource/model/pdfresource/events/pdfresourceondocformsave.class.php#L22

**Comment by Jako** [comment]
Could you try it with the following code in core/components/pdfresource/model/pdfresource/events/pdfresourceondocformsave.class.php

```
<?php
/**
 * @package pdfresource
 * @subpackage plugin
 */

class PDFResourceOnDocFormSave extends PDFResourcePlugin
{
    public function run()
    {
        // Generate the PDF if create_pdf template variable is checked
        /** @var modResource $resource */
        $resource = $this->modx->getOption('resource', $this->scriptProperties, '');
        /** @var modTemplateVar[] $tvs */
        $tvs = $resource->getTemplateVars();
        foreach ($tvs as $tv) {
            if ($tv->get('name') == $this->pdfresource->getOption('pdfTv', array(), 'create_pdf')) {
                $this->modx->switchContext($resource->context_key);
                $aliasPath = $resource->get('parent') ? preg_replace('#(\.[^./]*)$#', '/', $this->modx->makeUrl($resource->get('parent'))) : '';
                $this->modx->switchContext('mgr');

                $createPDF = intval($tv->getValue($resource->get('id')));
                if (!$createPDF && file_exists($this->pdfresource->getOption('pdfPath') . $aliasPath . $resource->get('alias') . '.pdf')) {
                    @unlink($this->pdfresource->getOption('pdfPath') . $aliasPath . $resource->get('alias') . '.pdf');
                } else {
                    $this->modx->invokeEvent('OnHandleRequest', array()); // call ClientConfig if installed
                    $this->modx->resource = &$resource;
                    $this->pdfresource->createPDF($resource, $aliasPath);
                }
            }
        }
    }
}
```

**Comment by pashajanx** [comment]
Your option works, thanks for the help.
Also need to remake and
core / components / pdfresource / model / pdfresource / events / pdfresourceonwebpageprerender.class.php

**Comment by pashajanx** [comment]
I noticed that it creates files incorrectly.
Snippet PDFResourceLink does not correctly give a link.
[Screenshot 1](https://prnt.sc/hjg500)
[Screenshot 2](https://prnt.sc/hjg5xm)

**Comment by Jako** [comment]
I don't know, why the PDF files are not placed in the right folder. The $aliaspath variable contains a trailing slash here in both event classes. The other bugs should be solved in 1.5.7.
